packages/cosmic-proto/src/address-hooks.js

```diff
@@ -0,0 +1,231 @@
+/**
+ * @module address-hooks
+ * @file This module provides functions for encoding and decoding address hooks
+ *   which are comprised of a bech32 base address with an HTTP query string, all
+ *   wrapped in a bech32 envelope.
+ * @example
+ *
+ *     import {
+ *       encodeAddressHook,
+ *       decodeAddressHook,
+ *     } from '@agoric/cosmic-proto/address-hooks.js';
+ *
+ *     const baseAddress = 'agoric1qqp0e5ys';
+ *     const query = { key: 'value', foo: ['bar', 'baz'] };
+ *
+ *     const addressHook = encodeAddressHook(baseAddress, query);
+ *     // 'agoric10rchqqplvehk70tzv9ezven0du7kyct6ye4k27faweskcat9qqqstnf2eq'
+ *
+ *     addressHook.startsWith('agoric10rch');
+ *     // true
+ *
+ *     const decoded = decodeAddressHook(addressHook);
+ *     // {
+ *     //   baseAddress: 'agoric1qqp0e5ys',
+ *     //   query: [Object: null prototype] { foo: [ 'bar', 'baz' ], key: 'value' }
+ *     // }
+ */
+
+/* eslint-disable no-bitwise */
+import { bech32 } from 'bech32';
+import queryString from 'query-string';
+
+// ADDRESS_HOOK_VERSION is the version of the address hook format used in this
+// module.
+const ADDRESS_HOOK_VERSION = 0;
+
+if ((ADDRESS_HOOK_VERSION & 0x0f) !== ADDRESS_HOOK_VERSION) {
+  throw Error(`ADDRESS_HOOK_VERSION ${ADDRESS_HOOK_VERSION} exceeds 0x0f`);
+}
+
+// AddressHookMagic is a magic byte prefix that identifies a hooked address.
+// Chosen to make bech32 address hooks that look like "agoric10rch..."
+const ADDRESS_HOOK_MAGIC = new Uint8Array([
+  0x78,
+  0xf1,
+  0x70 | ADDRESS_HOOK_VERSION,
+]);
+
+// The default maximum number of characters in a bech32-encoded hooked address.
+export const DEFAULT_HOOKED_ADDRESS_CHAR_LIMIT = 1024;
```

Comment:
This is API documentation, not an implementation detail comment, right?

```suggestion
/** The default maximum number of characters in a bech32-encoded hooked address. */
export const DEFAULT_HOOKED_ADDRESS_CHAR_LIMIT = 1024;
```

Reply:
re **Security Considerations**, I see:
> Advanced
> BIP173 enforces a limitation of 90 characters, if extend the LIMIT parameter beyond this, be aware that the **effectiveness of checksum decreases as the length increases.**

I don't suppose we're relying on a bech32 checksum for security beyond helping detect typos, though.

Reply:
That's right.  The checksum is just to avoid typo footguns, not providing any kind of authentication.